### ISSUE TITLE
Field comparison not ignoring synthetic members

### ISSUE THREAD

**Issue by petergphillips** [joel-costigliola/assertj-core#590]
Thanks for a great library.

We are using JaCoCo for test coverage.  This adds a private static field `$jacocoData` to classes under test, which is marked as synthetic (see My code uses reflection. Why does it fail when I execute it with JaCoCo? in http://eclemma.org/jacoco/trunk/doc/faq.html).

Assertj then fails the test when comparing classes field by field as the `$jacocoData` is different between the expected and actual.  So
```java
assertThat(eventClass).isEqualToIgnoringGivenFields(data, "types", "primaryMarkets");
```
then fails with the error:
```java
java.lang.AssertionError: 
Expecting value <[true, false, true, true]> in field <"$jacocoData"> but was <[true, true, true, true, true, true, true, true]> in <EventClass@2ef1972f>.
Comparison was performed on all fields but <["types", "primaryMarkets"]>
```

I can work round it by adding `$jacocoData` to the ignored fields, but this would be fail if JaCoCo changed to use a different name or we switched to an alternative coverage tool.  Ideally therefore instead assertj should ignore synthetic members as the JaCoCo FAQ suggests.

**Comment by joel-costigliola** [comment]
I agree, I'm just wandering if I should ignore all fields starting with `$` or not (I'm assuming there is no other way to detect synthetic fields).

**Comment by PascalSchumacher** [comment]
@joel-costigliola I guess `Field.isSynthetic()` can be used?

**Comment by joel-costigliola** [comment]
Yes ... I should have done my homework before asking ;-)

**Comment by petergphillips** [comment]
Agreed.   I was expecting something along the lines of `Objects.getDeclaredFieldsIncludingInherited` to exclude synthetic fields.  I could try create a patch if that would help...

**Comment by joel-costigliola** [comment]
@petergphillips  Feel free to do so but if you don't that's ok too.

**Comment by joel-costigliola** [comment]
In 2.x and soon in master, thanks

**Comment by LokiOfNZ** [comment]
Hi guys, this exact issue is also affecting us. Is there a plan to have the fix mentioned integrated into 3.4 also? And obviously 3.4 released?

**Comment by joel-costigliola** [comment]
Whatever goes in 2.4 will be in 3.4,  for the release you will have to wait,  I'm in holidays :-)  so probably in one month.
